### ISSUE TITLE
kill() broken in the shellcmd backend

### ISSUE THREAD

**Issue by GoogleCodeExporter** [uzh/gc3pie#458]
```
The `kill()` method in the ShellCmd backend is currently broken.  As
the following log snippet shows, it only tries to kill the single
process that it spawned (i.e., `/usr/bin/time ...`), while leaving all
of its children up and running::

        gc3.gc3libs: DEBUG: Connecting to host '130.60.24.184' as user 'gc3-user' via SSH (timeout 7s)...
        gc3.gc3libs: DEBUG: SshTransport running `kill 7232`...
        gc3.gc3libs: DEBUG: Executed command 'kill 7232' on host '130.60.24.184'; exit code: 0
        gc3.gc3libs: DEBUG: Deleting resource file for pid 7232

However, the process tree looks rather like this::

        gc3-user 15528  0.0  0.0   4300   348 ?        S    Jul09   0:00 /usr/bin/time -o /tmp/gc3libs.bqWIO5/.gc3pie_shellcmd/resource_usage.txt -f WallTime=%es?KernelTime=%Ss?UserTime=%Us?CPUUsage=%P?MaxResidentMemory=%MkB?AverageResidentMemory=%tkB?AverageTotalMemory=%KkB?AverageUnsharedMemory=%DkB?AverageUnsharedStack=%pkB?AverageSharedMemory=%XkB?PageSize=%ZB?MajorPageFaults=%F?MinorPageFaults=%R?Swaps=%W?ForcedSwitches=%c?WaitSwitches=%w?Inputs=%I?Outputs=%O?SocketReceived=%r?SocketSent=%s?Signals=%k?ReturnCode=%x /bin/sh -c  "./gcelljunction_wrapper.sh" "-x" "./tricellular_junctions_new" "--" "38"
        gc3-user 15529  0.0  0.0   4404   584 ?        S    Jul09   0:00  \_ /bin/sh -c  "./gcelljunction_wrapper.sh" "-x" "./tricellular_junctions_new" "--" "38"
        gc3-user 15530  0.0  0.0  11044  1412 ?        S    Jul09   0:00      \_ /bin/bash ./gcelljunction_wrapper.sh -x ./tricellular_junctions_new -- 38
        gc3-user 15535 53.7  2.1 1434392 175608 ?      Sl   Jul09 30838:45          \_ ./tricellular_junctions_new 38

Now this has also a very bad unintended consequence: since GC3Pie
deletes the resource file, the VM slot is then marked as free to run
other jobs, which it isn't  -- this is how we ended up having two
processes running in Tinri's VMs.

A good way to do the killing on Linux should be to kill all processes
that run in the same temporary directory::

    ec2ssh -n "gc3-user@$ip" "find /proc/*/cwd -maxdepth 0 -printf '%p %l\n' 2>/dev/null | egrep '${cwd}\$'" 2>/dev/null | \
        (while read proc wd; do
           pid=$(echo "$proc" | cut -d/ -f3)
           #echo "DEBUG: pid=$pid"
           #ec2ssh -n "gc3-user@$ip" "ps --no-headers u '$pid'" 2>/dev/null
           ec2ssh -n "gc3-user@$ip" "kill $pid" 2>/dev/null
        done)

I am not sure if the MacOSX `/proc` provides the same information,
though.

```

Original issue reported on code.google.com by `riccardo.murri@gmail.com` on 18 Aug 2014 at 12:48

**Comment by arcimboldo** [comment]
I am not sure this will work on MacOSX, but on linux you could select all the PIDs related to the session.

Given the PID of `time` (the only one GC3Pie knows about), you can recover the session ID with:

    ps -p 3989  -o sess --no-headers

and all the processes belonging to this session id with:

    ps -g <session-id> -o pid --no-headers

**Comment by arcimboldo** [comment]
kill $(ps -ax -o sess=,pid= | egrep "^[ \t]*$(ps -p %d  -o sess=)[ \t]")

this should work also on MacOSX. It basically finds the session group id of the parent process, then list all the processes and grep on the session id.

**Comment by arcimboldo** [comment]
Tested on a MacOSX 10.6 used as remote shellcmd resource, it seems to work.

**Comment by ehrenfeu** [comment]
Hi all,

regarding the proposed fix in ff2ec45 I'd like to mention that this unfortunately doesn't fix the problem for me: while it does kill all the child processes, it is killing way too aggressive - e.g. also my Python program itself (I'm running stuff on localhost) and also my shell wrapper scripts around it (they're used for automated testing).

I have tested this on Linux with both suggestions from the commit (the active one and also the commented out one) by integrating just these lines into my 2.4.2 installation. As the commit doesn't depend on other code (it's jut the `kill` command that is changed) I don't think this this will behave differently in master vs. 2.4.2 ...

In my opinion, this probably can't (easily) be solved in a one-shot approach, at least not in a cross-platform robust way. Maybe it is necessary to first retrieve a list of running processes with PIDs and PPIDs and from there assemble a list of process IDs to be killed, possibly extending the existing `lrms_jobid()` method?

One possible template for this approach could be taken from the `psutil` Python package's `Process.children()` method, see [`psutil/__init__.py#L801`](https://github.com/giampaolo/psutil/blob/caaacb30b013f2164e86548e1077002b0df541bf/psutil/__init__.py#L801).

One option to do it with shell code is provided in [StackExchange answer 85253](http://unix.stackexchange.com/a/124148/85253), although I'm not entirely sure about how platform-agnostic this is (it looks quite good though):
```bash
list_descendants () {
    local children=$(ps -o pid= --ppid "$1")
    for pid in $children ; do
        list_descendants "$pid"
    done
    echo "$children"
}

kill $(list_descendants $$)
```

Another approach could be to check if the target system has the [pslist](http://devel.ringlet.net/sysutils/pslist/) package available and then simply use the [`rkill`](https://manpages.debian.org/cgi-bin/man.cgi?query=rkill&apropos=0&sektion=0&manpath=Debian+8+jessie&format=html&locale=en) command - this would at least give the operators a way to provide a userspace tool to fix the issue.

Cheers
Niko

**Comment by ehrenfeu** [comment]
Hi @riccardomurri,

I did some further investigations, and at least in my case I think the main problem is that the `time` command does **not** notify its children when receiving a `HUP` signal, as illustrated here (notice how the parent-pid of the `sleep` process changes after `time` has been killed):

```bash
$ /usr/bin/time sleep 987654321 &
[1] 5555

$ ps -e -o uid,pid,ppid,cmd | grep '[ ]sleep 987654321'
 1000  5555 26982 /usr/bin/time sleep 987654321
 1000  5559  5555 sleep 987654321

$ kill 5555
[1]+  Terminated              /usr/bin/time sleep 987654321

$ ps -e -o uid,pid,ppid,cmd | grep '[ ]sleep 987654321'
 1000  5559  1819 sleep 987654321
```

In my situation it would be completely sufficient to send the signal to the (single) child process of `time`, as this one should properly deal with signals anyway. Even though my process is again spawning multiple children, it is properly cleaning up when receiving a `HUP` signal, i.e. also terminating its children (which is probably the only *reasonable* way to deal with signals, although there are good reasons why `time` behaves differently).

Long story short: for **me** an easy to implement workaround would be to figure out the PID of `time`'s single child and send the signal there, instead of killing the `time` process itself.

Feel free to move this conversation to the mailing list if you think this might produce valuable input on this topic...

Cheers
Niko

**Comment by ehrenfeu** [comment]
Hi,

I'd like to ask if there is any progress or plans on this - the `kill()` behavior is starting to be a real killer for me :frowning:

Cheers
~Niko

**Comment by riccardomurri** [comment]
I can commit to fixing by the end of this month.  Would that be OK with your deadlines?

**Comment by ehrenfeu** [comment]
End of may would be okay for me - this also depends a little on when you plan for doing a release containing this fix then. Personally I'm fine with installing from github master, but that's something I'd rather not put into official documentation :wink:

**Comment by ehrenfeu** [comment]
Hi guys, could you kindly give me an update on this? We're currently working on the next [HRM](https://github.com/aarpon/hrm) release, so I'd really need some fix for this issue...

**Comment by riccardomurri** [comment]
hi @ehrenfeu -- I'm working on it now, I hope to have some fix by tomorrow.  Sorry for the huge delay...

**Comment by riccardomurri** [comment]
I have pushed a (supposed) fix to branch "feature/shellcmd-rewrite" on
https://github.com/riccardomurri/gc3pie/
There is still a problem with TaskCollections, which I hope to solve ASAP
but if you're submitting a bunch of independent tasks it should be fine.

Can you please test it and tell me if it fixes the issue for you?

**Comment by ehrenfeu** [comment]
Hi @riccardomurri, I'll clone your repo and run my test scripts as soon as possible... Will keep you updated about the results, thanks!

**Comment by ehrenfeu** [comment]
Stupid question, but how do I install from your fork?

I cloned it, checked out the corresponding branch and ran the following command:
```
python install.py --feature=none --development --no-gc3apps
```
but the resulting installation is based on the uzh repository, not yours...

**Comment by riccardomurri** [comment]
1. install gc3pie as usual
2. clone my repo
3. run `pip install .` in the cloned repo with the gc3pie virtualenv active

**Comment by ehrenfeu** [comment]
This didn't work, the files in the virtualenv were still the same after running the pip install.
Never mind, I followed the steps for the manual installation and here it seems to be fine.

However, when starting my application I'm running into trouble directly as the resource attributes seem to have changed. My code looks like this:
```python
for resource in engine.get_resources():
    resourcedir = os.path.expandvars(resource.cfg_resourcedir)
```
which leads to this error:
```
AttributeError: 'ShellcmdLrms' object has no attribute 'cfg_resourcedir'
```
Any hint?

**Comment by ehrenfeu** [comment]
Found it, its `_resourcedir_raw` now - correct?

**Comment by riccardomurri** [comment]
> Found it, its `_resourcedir_raw` now - correct?

Yes.  But why are you using it and not `.resourcedir`?  (The reason it was renamed was to make it clear that it's an internal attribute, to be expanded by a call to the remote resource, not a final value for other clients consumption...)

**Comment by riccardomurri** [comment]
There were still a couple of errors, which I just corrected.  You might need to clone the repo again as I used `push -f`.

It now works for me, as soon as you confirm I'll merge into master.  (Pleas let me know if you don't have time to test, though.)

**Comment by ehrenfeu** [comment]
I can confirm the job killing seems to work as expected now. However

- the `.resourcedir` doesn't work for me, even after pulling in the most recent changes - I still receive an `AttributeError`
- **much worse**: regular job processing seems to be broken now, no job ever reaches the `TERMINATED` state, instead they all remain in `TERMINATING` - is this to be expected? I wouldn't think so...

**Comment by riccardomurri** [comment]
Can you please do::

    git rev-parse HEAD

on the GC3Pie source repository that you're using right now?

**Comment by ehrenfeu** [comment]
Sure, it's the current HEAD of your topic branch:
f1fed9a062bd4b36dd1c5107005e80916e2ceb00

**Comment by riccardomurri** [comment]
Hmmmm... no, the topic branch should be on http://github.com/riccardomurri/gc3pie/ not .../uzh/gcpie/

Can you please try cloning again now? HEAD is currently at https://github.com/uzh/gc3pie/pull/626/commits/c5482c4e53eb3e8b89dcd1465dc4987f358c4309

**Comment by riccardomurri** [comment]
Sorry, HEAD is 333ebe4 -- there was a trivial typo that I just fixed.

**Comment by riccardomurri** [comment]
Re: the "resource dir" problem: the exact spelling of the attribute is `.resource_dir` (with an underscore), can you please check?

**Comment by ehrenfeu** [comment]
Sure, riccardomurri/gc3pie@f1fed9a was on your repo, not on uzh (as you can see with your own SHA-1 links, github turns them into references to the repo where *this* issue belongs to unless they're explicitly prefixed) - before you did another force-push (its now riccardomurri/gc3pie@7ed7f12).

I'll try again with the most recent changes.

**Comment by riccardomurri** [comment]
Re: HEAD, the top commit should be riccardomurri/gc3pie@333ebe4 not 7ed7f12 ...

**Comment by ehrenfeu** [comment]
Yes, I was just explaining what corresponds to f1fed9a in the new tree, which was what I had been using with my previous tests. Hence the:
> I'll try again with the most recent changes

Using riccardomurri/gc3pie@333ebe4 I get a new stack trace:
```
  File "/home/noenc_ehrenfeu/usr/packages/IMCF/snijder/src/snijder/spooler.py", line 212, in resource_dirs_clean
    resourcedir = os.path.expandvars(resource.resource_dir)
  File "/home/ehrenfeu/gc3pie/src/gc3libs/backends/shellcmd.py", line 649, in resource_dir
    self._init_resource_dir()
  File "/home/ehrenfeu/gc3pie/src/gc3libs/backends/shellcmd.py", line 659, in _init_resource_dir
    'echo %s' % sh_quote_unsafe(self._resourcedir_raw))
  File "/home/ehrenfeu/gc3pie/src/gc3libs/backends/transport.py", line 941, in execute_command
    "`Transport.execute_command()` called" \
AssertionError: `Transport.execute_command()` called on closed (or not yet opened) `Transport` instance.
```

**Comment by riccardomurri** [comment]
Looks like you're calling `.resource_dir` before the backend is initialized.

Does the error go away if you call `resource.get_resource_status()` before getting `resource.resource_dir` ?
(You just need to call it once.)

**Comment by riccardomurri** [comment]
Re: `.resource_dir` should be fixed in 2a887b9

**Comment by ehrenfeu** [comment]
Okay, `.resource_dir` is now working as expected.

However, the `TERMINATING` issue persists - my jobs never reach the `TERMINATED` state (unlike with the last official gc3pie release).

**Comment by riccardomurri** [comment]
Okay, .resource_dir is now working as expected.


Good 😀

However, the TERMINATING issue persists - my jobs never reach the TERMINATED
state (unlike with the last official gc3pie release).

I would need to know more about how you are managing the jobs then. It
doesn't happen here so you probably are applying a different code pattern...

**Comment by riccardomurri** [comment]
Let's do it like this: I merge the branch into master, and we debug issues with TERMINATED/TERMINATING from there -- it looks like it's a different issue altogether.

**Comment by ehrenfeu** [comment]
Sorry for the delay, I'm still recovering from a pretty bad virus incident (personal health, not computer problem :wink:).
If you prefer, go ahead and merge into master.

Should I open a new issue then? If yes, then where - here in the uzh repo or in you personal one?

In the meantime I did some more debugging and found out the problem only occurs with jobs that have the `outputs` parameter set to non-empty. Increasing the loglevel to maximum results in the following messages:
```
gc3.gc3libs: DEBUG: About to update state of application: HuDeconApp@7fb684473550 (currently: SUBMITTED)
gc3.gc3libs: DEBUG: Executed local command 'ps -p 25531 -o state=', got exit status: 1
gc3.gc3libs: DEBUG: Process with PID 25531 not found, assuming task HuDeconApp@7fb684473550 has finished running.
gc3.gc3libs: DEBUG: Calling state-transition handler 'terminating' on HuDeconApp@7fb684473550 ...
gc3.gc3libs: DEBUG: Updating job info file for pid 25531
gc3.gc3libs: DEBUG: Reading resource utilization from wrapper file `/data/snijder_spooldir/gc3libs.XEGCfR/.gc3pie_shellcmd/resource_usage.txt` for task HuDeconApp@7fb684473550 ...
gc3.gc3libs: DEBUG: Deleting job info file for pid 25531 ...
gc3.gc3libs: DEBUG: Removing /data/gc3_resourcedir/shellcmd.d/25531
gc3.gc3libs: DEBUG: Recovered resource information from files in /data/gc3_resourcedir/shellcmd.d: total nr. of cores: 2, requested by jobs: 0; available memory: 2000MB, requested by jobs: 0MB.
gc3.gc3libs: DEBUG: Ignored error in fetching output of task 'HuDeconApp@7fb684473550': AttributeError: 'ShellcmdLrms' object has no attribute '_make_remote_and_local_path_pair'
gc3.gc3libs: DEBUG: (Original traceback follows.)
Traceback (most recent call last):
  File "/home/ehrenfeu/gc3pie/src/gc3libs/core.py", line 1805, in progress
    changed_only=self.retrieve_changed_only)
  File "/home/ehrenfeu/gc3pie/src/gc3libs/core.py", line 607, in fetch_output
    app, download_dir, overwrite, changed_only, **extra_args)
  File "/home/ehrenfeu/gc3pie/src/gc3libs/core.py", line 675, in __fetch_output_application
    raise ex
AttributeError: 'ShellcmdLrms' object has no attribute '_make_remote_and_local_path_pair'
```

**Comment by riccardomurri** [comment]
Hi Niko, the traceback is really telling: I had removed this function since
I could find no references to it in the current "master".

Anyway, I'm traveling over the weekend, so no action until Tuesday.

Take your time to recover, there's no hurry 😃

Ciao,
R

**Comment by riccardomurri** [comment]
Merged into branch "master" at commit f5bb2f7 -- the issue with the "shellcmd" backend not properly cleaning up jobs should now be fixed for good.

(I'm opening a separate issue re: "'ShellcmdLrms' object has no attribute '_make_remote_and_local_path_pair'")
